### ISSUE TITLE
Improve accuracy

### ISSUE THREAD

**Issue by xsacha** [mahfujur1/Face-mask-Classification-PyTorch#1]
You should be able to run a face detection + alignment first to limit the input variation for the classification model.
I'd suggest retinaface or similar.

**Comment by mahfujur1** [comment]
@xsacha Thank you so much. In this repo, I am trying to compare the results using two popular transfer learning strategy.  It would be appreciable if you want to contribute on this project.

**Comment by xsacha** [comment]
Ok. Just thought I'd help out a fellow QUT student. ;)

You should have a test dataset as well because the high accuracy on val is misleading (being from the same set). It may not show the benefits from the transfer learning you are expecting.

Here's another dataset if you are having trouble finding some too:
https://drive.google.com/file/d/1QspxOJMDf_rAWVV7AU_Nc0rjo1_EPEDW/view (WIDER + MAFA)

**Comment by mahfujur1** [comment]
@xsacha Thank you so much for providing the dataset that would be really helpful for mask detection. It's really nice to talk with a fellow from Brisbane.

The model accuracy on 1376 test images is 98.91% (NOT BAD AT ALL).

I used Resnet-18 pre-trained network on ImageNet dataset. Certainly, it shows the benefits of the transfer learning because I used the pre-trained network on a smaller dataset ( mask dataset) which has only 2757 training images and 1376 test images. 

Here, the val data is the test data. I did not split the training images as train and val. Rather I used test data (val data) to check the model accuracy  per epoch on test data.

Thank you so much. Please let me know if you have any further quires.

**Comment by xsacha** [comment]
Thanks.

I have pre-processed the dataset with our inhouse retinaface to improve the accuracy. This makes it easier for the pre-trained (imagenet) resnet to distinguish between the features.

See:
![image](https://user-images.githubusercontent.com/61218/82510630-6ad01e80-9b4e-11ea-9df0-e9bf59bf6213.png)

![image](https://user-images.githubusercontent.com/61218/82510435-dbc30680-9b4d-11ea-8b4c-4fbddd637a73.png)

It also makes it easier to spot duplicates and images that contain faces without masks (and skew the data).

Some images such as train/mask/731 are ambiguous. She is wearing a mask, but the mouth is cut out of it. train/without_mask/366, her nose and mouth is fully covered.

The augmented validation images are also fairly useless if you have test time augmentation anyway. I found the real sample size of val is about 500 images. Since the masks are generated, this could be any size you want :)

**Comment by xsacha** [comment]
Accuracy and train time improvements:

On finetune:
By epoch **2** it had 98.98% accuracy on val as compared to 98.91% on epoch **24** without face detection.
By epoch 6, it had 99.78% accuracy on val and was effectively fully trained.
Best epoch:

Epoch 23/24
train Loss: 0.0050 Acc: 0.9985
val Loss: 0.0150 Acc: 0.9985

Best val Acc: 0.998537

On Convnet:
Epoch 5/24
train Loss: 0.3468 Acc: 0.8756
val Loss: 0.0565 Acc: 0.9854

Best val Acc: 0.985369


Model accuracy on 1367 test images: 99.85%
Category : Accuracy
mask  : 99.7
without_mask : 100.0
actual/pred     mask      without_mask
mask            683       2
without_mask    0         682

**Comment by xsacha** [comment]
As a follow up, I found the 2 images that failed validation and it turns out they in fact did not have masks. Removing these two images pushed test accuracy to 100%.

I removed the normalisation (use in-built float normalisation instead) and I get 100% after 7th epoch.

Model accuracy on 1367 test images: 100.00%
Category : Accuracy
mask : 100.0
without_mask : 100.0
actual/pred mask without_mask
mask 683 0
without_mask 0 684

**Comment by mahfujur1** [comment]
@xsacha Thank you so much. Really impressive.
